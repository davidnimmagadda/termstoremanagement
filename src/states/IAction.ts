/**
 * We are following FSA (Flux Standard Action) compliant actions, which is mentioned in Redux documentations.
 * This Flux is not a typo :)
 */
export interface IAction {
  /**
   * We use strings to represent action types.
   * Each page should have separated definitions for all allowed types.
   *
   * For example:
   *
   * type enum ActionTypes = 'ADD_ITEM' | 'REMOVE_ITEM';
   */
  type: string;

  /**
   * The optional payload can be anything holds, well, the payload.
   * If `error` is `true`, the payload should be an error object.
   */
  // tslint:disable-next-line:no-any
  payload?: any;

  /**
   * A `true` value means the action has an error.
   * Any other values, including `undefined` and `null`, means the action is good.
   */
  error?: boolean;

  /**
   * The optional meta can be anything holds extra information that is not part of the payload.
   */
  // tslint:disable-next-line:no-any
  meta?: any;
}
