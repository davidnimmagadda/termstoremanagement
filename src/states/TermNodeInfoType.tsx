/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

enum TermNodeInfoType {
  FOLDER = 'folder',
  FILE = 'file'
}

export default TermNodeInfoType;
