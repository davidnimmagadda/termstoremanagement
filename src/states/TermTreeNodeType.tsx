/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

enum TermTreeNodeType {
  EMPTY = '',
  TERM_STORE = 'TERM_STORE',
  TERM_GROUP = 'TERM_GROUP',
  TERM_SET = 'TERM_SET',
  TERM = 'TERM'
}

export default TermTreeNodeType;
