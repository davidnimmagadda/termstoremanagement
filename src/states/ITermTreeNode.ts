/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

import TermTreeNodeType from './TermTreeNodeType';
import TermNodeInfoType from './TermNodeInfoType';

export interface ITermTreeNode {
  nodeType: TermTreeNodeType;
  level: number;
  childrenUri: string;
  isOpen?: boolean;
  info: ITermTreeNodeAPIResponse;
}

export interface ITermTreeNodeAPIResponse {
  id: string;
  type: TermNodeInfoType;
  name: string;
}
