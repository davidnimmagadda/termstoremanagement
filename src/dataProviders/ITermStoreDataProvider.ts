/**
 * @copyright Microsoft Corporation. All rights reserved.
 */
import { ITermTreeNodeAPIResponse } from '../states/ITermTreeNode';

export interface ITermStoreDataProvider {
  getTermStore(): Promise<ITermTreeNodeAPIResponse>;

  getNodeChildren(uri: string): Promise<ITermTreeNodeAPIResponse[]>;
}
