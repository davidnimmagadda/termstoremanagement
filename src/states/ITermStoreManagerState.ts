/**
 * @copyright Microsoft Corporation. All rights reserved.
 */
import { ITermTreeNode } from './ITermTreeNode';

export interface ITermStoreManagerState {
  selectedTermTreeNode: ITermTreeNode;
}
