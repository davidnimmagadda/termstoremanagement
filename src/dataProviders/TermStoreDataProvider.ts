/**
 * @copyright Microsoft Corporation. All rights reserved.
 */
import { ITermStoreDataProvider } from './ITermStoreDataProvider';
import { ITermTreeNodeAPIResponse } from '../states/ITermTreeNode';
import { ITermStoreEndpoint } from './endpoints/ITermStoreEndpoint';
import { MockTermStoreEndpoint } from './endpoints/MockTermStoreEndpoint';
// ES6 Module implementation of a singleton
const _endpoint: ITermStoreEndpoint = MockTermStoreEndpoint;

function getTermStore(): Promise<ITermTreeNodeAPIResponse> {
  return _endpoint.getTermStore();
}

function getNodeChildren(uri: string): Promise<ITermTreeNodeAPIResponse[]> {
  return _endpoint.getNodeChildren(uri);
}

// tslint:disable-next-line:variable-name
export const TermStoreDataProvider: ITermStoreDataProvider = {
  getTermStore,
  getNodeChildren
};
