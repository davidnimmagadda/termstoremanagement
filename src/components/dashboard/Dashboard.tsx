/**
 * @copyright Microsoft Corporation. All rights reserved.
 */
import * as React from "react";
import { Spinner, SpinnerSize } from "office-ui-fabric-react";

import TreeNode from "../tree/TreeNode";
import * as utils from "../tree/TreeUtils";
import {
  ITermTreeNode,
  ITermTreeNodeAPIResponse
} from "../../states/ITermTreeNode";
import { TermStoreDataProvider } from "../../dataProviders/TermStoreDataProvider";
import actionTypes from "../../actions/TermStoreAdminActionTypes";
import "../../TermStoreAdminCenterPage.css";
import TermTreeNodeType from "../../states/TermTreeNodeType";
import selectedTermTreeNodeReducer, {
  initialState
} from "../../reducers/selectedTermTreeNodeReducer";

export function Dashboard(): JSX.Element {
  const [selectedTermTreeNode, dispatch] = React.useReducer(
    selectedTermTreeNodeReducer,
    initialState
  );

  React.useEffect(() => {
    loadTermStore();
  }, []);

  async function loadTermStore(): Promise<void> {
    // TODO danimmag: Will be replaced with actual API call using odata client. This is a mock API
    const response: ITermTreeNodeAPIResponse = await TermStoreDataProvider.getTermStore();
    dispatch({
      type: actionTypes.LOAD_TERM_STORE_SUCCESS,
      payload: response
    });
  }

  function getRightPane(): JSX.Element {
    return (
      <div style={{ marginLeft: "20px" }}>
        <h3>{selectedTermTreeNode.nodeType}</h3>
      </div>
    );
  }

  function onNodeSelect(node: ITermTreeNode): void {
    dispatch({
      type: actionTypes.ON_TREE_NODE_SELECT,
      payload: node
    });
  }

  return (
    <>
      <div className="headerContainer">
        <h1>Term Store Admin Center</h1>
      </div>
      <div className="dashboard">
        {selectedTermTreeNode.nodeType === TermTreeNodeType.EMPTY ? (
          <div className="loaderContainer">
            <Spinner size={SpinnerSize.large} />
          </div>
        ) : (
          <>
            <div className="treeContainer">
              <TreeNode
                currentNode={selectedTermTreeNode}
                selectedTermTreeNode={selectedTermTreeNode}
                show={true}
                onNodeSelect={onNodeSelect}
                loadNodeChildren={utils.loadNodeChildren}
              />
            </div>
            <div className="rightPane">{getRightPane()}</div>
          </>
        )}
      </div>
    </>
  );
}

export default Dashboard;
