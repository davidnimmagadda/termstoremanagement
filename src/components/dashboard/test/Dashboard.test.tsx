/**
 * @copyright Microsoft Corporation. All rights reserved.
 */

import * as React from "react";
import { mount, ReactWrapper } from "enzyme";
import { act } from "react-dom/test-utils";
import TestRenderer, { create, ReactTestRenderer } from "react-test-renderer";

import { Dashboard } from "../Dashboard";
import { ITermTreeNodeAPIResponse } from "../../../states/ITermTreeNode";
import { TermStoreDataProvider } from "../../../dataProviders/TermStoreDataProvider";
import TermTreeNodeType from "../../../states/TermTreeNodeType";
import TermNodeInfoType from "../../../states/TermNodeInfoType";
import { Spinner } from "office-ui-fabric-react";

describe("Term Store Admin Center Page", () => {
  describe("Dashboard", () => {
    let wrapper: ReactWrapper;
    // tslint:disable-next-line:no-any
    let getTermStoreMock: jest.SpyInstance<
      Promise<ITermTreeNodeAPIResponse>,
      []
    >;

    beforeEach(() => {
      getTermStoreMock = jest.spyOn(TermStoreDataProvider, "getTermStore");
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    const termSetSelectedNode: ITermTreeNodeAPIResponse = {
      id: "23234",
      type: TermNodeInfoType.FOLDER,
      name: "test"
    };

    it("should render same as previous", async () => {
      getTermStoreMock.mockImplementation(() =>
        Promise.resolve(termSetSelectedNode)
      );
      let tree: ReactTestRenderer = (undefined as unknown) as ReactTestRenderer;
      await TestRenderer.act(async () => {
        tree = create(<Dashboard />);
      });
      expect(tree).toMatchSnapshot();
    });

    it("should render loader when nodeType is empty and load function should be called", () => {
      getTermStoreMock.mockImplementation(() =>
        Promise.resolve({} as ITermTreeNodeAPIResponse)
      );
      act(() => {
        wrapper = mount(<Dashboard />);
      });
      expect(TermStoreDataProvider.getTermStore).toHaveBeenCalled();
    });

    it("should display corresponding nodetype in right pane when not empty, api should not be called", async () => {
      getTermStoreMock.mockImplementation(() =>
        Promise.resolve(termSetSelectedNode)
      );
      await act(async () => {
        wrapper = mount(<Dashboard />);
      });
      wrapper.update();
      expect(TermStoreDataProvider.getTermStore).toHaveBeenCalled();
      expect(wrapper.find("h3").text()).toEqual(TermTreeNodeType.TERM_STORE);
    });

    it("should display loader first and after api is called back, will display, tree and right pane", async () => {
      jest.useFakeTimers();
      getTermStoreMock.mockImplementation(() => {
        setTimeout(() => {}, 1000);
        return Promise.resolve(termSetSelectedNode);
      });
      await act(async () => {
        wrapper = mount(<Dashboard />);
      });
      expect(wrapper.find(Spinner)).toHaveLength(1);
      jest.runAllTimers();
      wrapper.update();
      expect(wrapper.find(Spinner)).toHaveLength(0);
      expect(TermStoreDataProvider.getTermStore).toHaveBeenCalled();
      expect(wrapper.find("h3").text()).toEqual(TermTreeNodeType.TERM_STORE);
    });
  });
});
