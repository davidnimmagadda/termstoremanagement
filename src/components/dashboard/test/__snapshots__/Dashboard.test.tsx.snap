// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Term Store Admin Center Page Dashboard should render same as previous 1`] = `
Array [
  <div
    className="headerContainer"
  >
    <h1>
      Term Store Admin Center
    </h1>
  </div>,
  <div
    className="dashboard"
  >
    <div
      className="treeContainer"
    >
      <div
        className="treeNode"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "display": "flex",
            "paddingLeft": 0,
          }
        }
      >
        <span
          data-automation-id="treeNodeToggleChevron"
          onClick={[Function]}
          role="button"
        >
          <i
            aria-hidden={true}
            className="treeNodeIcon root-51"
            data-icon-name="ChevronRightMed"
            role="presentation"
          >
            
          </i>
        </span>
        <i
          aria-hidden={true}
          className="root-53"
          data-icon-name="FabricFolderFill"
          role="presentation"
          style={
            Object {
              "marginRight": 5,
            }
          }
        >
          
        </i>
        <span
          data-automation-id="treeNodeDisplayName"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
        >
          test
        </span>
        <button
          className="ms-Button ms-Button--icon root-54"
          data-is-focusable={true}
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          style={
            Object {
              "visibility": "hidden",
            }
          }
          type="button"
        >
          <div
            className="ms-Button-flexContainer flexContainer-55"
          >
            <i
              aria-hidden={true}
              className="ms-Button-icon icon-62"
              data-icon-name="MoreVertical"
              role="presentation"
            >
              
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      className="rightPane"
    >
      <div
        style={
          Object {
            "marginLeft": "20px",
          }
        }
      >
        <h3>
          TERM_STORE
        </h3>
      </div>
    </div>
  </div>,
]
`;
